site/content/contribute/server/plugins.md

```diff
@@ -0,0 +1,42 @@
+
+The JavaScript bundle is a webpack-built collection of JavaScript code that will be run on the Mattermost web/desktop apps. When a plugin is enabled, the client is notified and it makes a request to add the JS bundle to the document. The plugin's client code then registers itself and components with the Mattermost client through the client's [plugin registry](https://github.com/mattermost/mattermost-webapp/blob/master/plugins/registry.js). The registry contains many methods for registering different components and callbacks. These are all stored as part of the app's [plugin reducer](https://github.com/mattermost/mattermost-webapp/blob/master/reducers/plugins/index.js). The [Pluggable](https://github.com/mattermost/mattermost-webapp/tree/master/plugins/pluggable) component is then inserted into various places in the app, allowing plugins to insert components into these locations in the UI. In some special cases, the Pluggable component is not used and we instead implement the plugs manually. More information is available on the [web app side of plugins in the author documenation](/extend/plugins/webapp).
+
+In the future there will be another component of plugins for the mobile apps, likely a React Native bundle.
+
+All these different components of a plugin are compressed into a .tar.gz bundle. Installing a plugin is the process of uploading this bundle to the Mattermost server (via the UI, REST API or CLI). The server then unpacks the bundle, performs some validation and extracts it into the configured directory for storing installed plugins. Installed plugins are not yet running. To start a plugin it must be enabled (again via the UI, REST API or CLI). Once it is enabled, the server will then start the server process and prepare the web app bundle for serving to the client. Plugin settings, configuration and enabled/disabled status are managed by the Mattermost `config.json` using a [PluginSettings](https://godoc.org/github.com/mattermost/mattermost-server/model#PluginSettings) struct.
+
+Check out the [`plugin` package](https://github.com/mattermost/mattermost-server/tree/master/plugin) and the [plugin_* files in the `app` package](https://github.com/mattermost/mattermost-server/tree/master/app) for the code, and [mattermost-plugin-demo](https://github.com/mattermost/mattermost-plugin-demo) for an example plugin.
+
+## Adding an API 
+
+To add a plugin API you need to add the signature of your new method to the [API interface](https://github.com/mattermost/mattermost-server/blob/master/plugin/api.go). You then need to implement the API in the [plugin_api.go](https://github.com/mattermost/mattermost-server/blob/master/app/plugin_api.go) of the `app` package. Finally, you would need to run `make pluginapi` to generate the RPC glue code needed for your new API.
```

Comment:
Are any (unit) tests added?

Reply:
Testing of the wireframe is covered by the auto-generated code. For more complex API implementations we'll want separate tests but that'll be the special case so we'll ask for it case-by-case generally